### PR TITLE
Issue 11: Coverage of message specifications

### LINKED ISSUE #11: Test coverage of message specifications
It would be helpful to determine the test coverage of the message specifications. The validation tool (#7) would need to analyze how many paths of the message graph are covered by the existing test data.

----
Implementation Plan

MessageValue
- [x] Add attribute (property) `covered_links: List[Link]`
- [x] Add `_next_link(field_name) -> Link` method which returns the next valid link 
- [x] Call `next_link` inside `next_field` and store the last returned link in `covered_links`
- [x] Add `get_sdu_messages -> List[MessageValue]` which returns the list of all messages contained in all fields of type `OpaqueValue` and `ArrayValue` 
- [x] Add tests

Validation Tool

- [x] create dict to store the coverage of all involved messages (models) - total coverage
- [x] method to update the total coverage on every iteration
- [x] generate output
- [x] tests

Output

Short Output: 
```
------------------------------------------------------------------------------
                     RecordFlux Validation Coverage Report
Directory: .
------------------------------------------------------------------------------
File                                       Links    Used   Coverage  
------------------------------------------------------------------------------
ethernet.rflx                                100     100   100%
in_ethernet.rflx                              34      34   100%
ipv4.rflx                                    206      97    47%   
------------------------------------------------------------------------------
TOTAL                                        340     231    68%
------------------------------------------------------------------------------
```

Full Output:
`<file_name>:<line>:<col>: error: missing link <link_source_field> -> <link_target_field>: <link_condition>`

### PR DESCRIPTION
**Pull request by rssen**
# Description

This PR adds functionality analysing the coverage of message specifications to the validation tool. 
Specify the `-c` or `--coverage` flag to get information about the link coverage reached by the messages in the valid and invalid directory on the specification files. Specify `--target-coverage=x.x` to set a threshold. If the target coverage is missed, the validation tool will exit with exit code 1.

The pylint test is failing for an unknown reason. It is executed successfully on my local machine.

Closes #11

# Review Checklist

### Implementation

- [x] Does this code change accomplish what it is supposed to do?
- [x] Can this solution be simplified?
- [x] Does this change add unwanted compile-time or run-time dependencies?
- [x] Was a framework, API, library, service used that should not be used?
- [x] Was a framework, API, library, service not used that could improve the solution?
- [x] Is the code at the right abstraction level?
- [x] Is the code modular enough?
- [x] Would you have solved the problem in a different way that is substantially better in terms of the code’s maintainability, readability, performance, security?
- [x] Does similar functionality already exist in the codebase? If so, why isn’t this functionality reused?
- [x] Are there any best practices, design patterns or language-specific patterns that could substantially improve this code?

### Logic Errors and Bugs

- [x] Can you think of any use case in which the code does not behave as intended?
- [x] Can you think of any inputs or external events that could break the code?

### Error Handling and Logging

- [x] Is error handling done the correct way?
- [x] Should any logging or debugging information be added or removed?
- [x] Are error messages user-friendly?
- [x] Are there enough log events and are they written in a way that allows for easy debugging?
- [x] Is the style of error / logging messages consistent?

### Dependencies

- [x] If this change requires updates outside of the code, like updating the documentation, configuration, readme files, was this done?
- [x] Might this change have any ramifications for other parts of the system, or backward compatibility?

### Security and Data Privacy

- [x] Does this code open the software up for security vulnerabilities?
- [x] Does this code change reveal some secret information like keys, passwords, or usernames (e.g., checked-in credentials)?

### Performance

- [x] Do you think this code change will impact system performance in a negative way?
- [x] Do you see any potential to improve the performance of the code?

### Usability

- [x] Is the proposed solution well designed from a usability perspective?
- [x] Is the API well documented?
- [x] Is the API/UI intuitive to use?

### Testing

- [x] Does it have enough automated tests (unit/integration/system tests)?
- [x] Are there some test cases, input or edge cases that should be tested in addition?

### Readability

- [x] Was the code easy to understand?
- [x] Can the readability of the code be improved by smaller methods?
- [x] Can the readability of the code be improved by different function/method or variable names?
- [x] Is the code located in the right file/folder/package?
- [x] Do you think certain methods should be restructured to have a more intuitive control flow?
- [x] Is the data flow understandable?
- [x] Are there redundant comments?
- [x] Could some comments convey the message better?
- [x] Would additional comments make the code more understandable?
- [x] Could some comments be removed by making the code itself more readable?
- [x] Is there any commented out code?

### Experts Opinion

- [x] Do you think a specific expert, like a security expert or a usability expert, should look over the code before it can be committed?
- [x] Will this code change impact different teams?
- [x] Should they have a say on the change as well?

### CODE REVIEW

File: tools/validate_spec.py [hunk 1]
```diff
@@ -226,10 +323,67 @@ def __exit__(
...
+            print(f"{file : ^80}")
+            print("-" * 80)
+            uncovered_links = total_coverage.file_uncovered_links(file)
+            if len(uncovered_links) == 0:
+                print(f"{'all links covered':^80}")
+                continue
+            for link in uncovered_links:
+                print(
+                    f"{str(link.location) if link.location is not None else '':<17}"
+                    f": missing link {link.source.name:^25} -> {link.target.name:^20}"
```
Comment (treiher) [inline tools/validate_spec.py:385]: I would rather like to have a "None" instead of nothing for missing locations. Furthermore, there should be no missing link locations in the future.

File: tools/validate_spec.py [hunk 2]
```diff
@@ -122,10 +146,13 @@ def validate(
...
             for path in directory:
                 validation_result = _validate_message(path, is_valid_directory, message_value)
+                if coverage_info is not None:
+                    coverage_info.update(validation_result.parser_result)
                 output_writer.write_result(validation_result)
                 if not validation_result.validation_success and abort_on_error:
                     raise ValidationError(f"aborted: message {path} was classified incorrectly")
-
+        if coverage_info is not None:
+            output_writer.write_coverage(coverage_info)
```
Comment (treiher) [inline tools/validate_spec.py:155]: The `is not None` checks could be done inside the method. The readability just gets worse with these checks.

File: tests/test_specs.py [hunk 3]
```diff
@@ -43,6 +43,7 @@ def test_validate_spec(spec: str) -> None:
                     directory_valid,
                     json_output=None,
                     abort_on_error=False,
+                    coverage_info=None,
```
Comment (treiher) [inline tests/test_specs.py:46]: The coverage calculation should be enabled for this test.

**Review by treiher — CHANGES_REQUESTED** [review]

File: tools/validate_spec.py (hunk 2 above)
Comment (rssen) [reply tools/validate_spec.py]: Done in ad03d50

File: tools/validate_spec.py [hunk 4]
```diff
@@ -226,10 +323,67 @@ def __exit__(
...
+                err.append(
+                    f"missed target coverage of {self.target_coverage:.2%}, "
+                    f"reached {self.reached_coverage:.2%}"
+                )
+            if len(err.messages) > 0:
+                raise err
+
+    def write_coverage(self, coverage_info: CoverageInformation) -> None:
+        self.__write_coverage_short(coverage_info)
+        self.__write_coverage_full(coverage_info)
```
[the remark this answers is not in the document]
Comment (rssen) [reply tools/validate_spec.py]: Done in ad03d50

File: tools/validate_spec.py [hunk 5]
```diff
@@ -226,10 +323,67 @@ def __exit__(
...
+        self.reached_coverage = coverage_info.total_covered_links / coverage_info.total_links
+        self.target_coverage = coverage_info.target_coverage
```
[the remark this answers is not in the document]
Comment (rssen) [reply tools/validate_spec.py]: Moved the output Methods into `CoverageInformation` in ad03d50

File: tools/validate_spec.py [hunk 6]
```diff
@@ -99,6 +121,7 @@ def cli(argv: List[str]) -> Union[int, str]:
             args.directory_valid,
             args.json_output,
             args.abort_on_error,
+            coverage_info,
         )
```
[the remark this answers is not in the document]
Comment (rssen) [reply tools/validate_spec.py:121]: Changed in ad03d50

File: tools/validate_spec.py [hunk 7]
```diff
@@ -161,6 +188,74 @@ def _validate_message(
     )
 
 
+class CoverageInformation:
+    def __init__(self, packages: List[Package], target_coverage: float) -> None:
+        self._total_message_coverage: Dict[str, Dict[Link, bool]] = {}
+        self.spec_files: Dict[str, List[str]] = {}
+        self.target_coverage = target_coverage
```
[the remark this answers is not in the document]
Comment (rssen) [reply tools/validate_spec.py]: It is now only used in `validate`. ad03d50

[the remark this answers is not in the document]
Comment (rssen) [reply tools/validate_spec.py]: Changed in ad03d50

File: tools/validate_spec.py [hunk 8]
```diff
@@ -161,6 +188,74 @@ def _validate_message(
...
+        )
+
+    def file_covered_links(self, file_name: str) -> int:
+        assert file_name in self.spec_files
+        return sum(
+            [
+                list(self._total_message_coverage[message].values()).count(True)
+                for message in self.spec_files[file_name]
+            ]
+        )
```
[the remark this answers is not in the document]
Comment (rssen) [reply tools/validate_spec.py]: Done in 398135a

File: tools/validate_spec.py [hunk 9]
```diff
@@ -161,6 +188,74 @@ def _validate_message(
...
+                if file_name in self.spec_files:
+                    self.spec_files[file_name].append(str(message.identifier))
+                else:
+                    self.spec_files[file_name] = [str(message.identifier)]
+
+    def update(self, message_value: MessageValue) -> None:
+        messages = message_value.inner_messages() + [message_value]
+        for message in messages:
+            for link in message.path:
+                self._total_message_coverage[str(message.identifier)][link] = True
```
[the remark this answers is not in the document]
Comment (rssen) [reply tools/validate_spec.py]: The `inner_messages()` method of MessageValue already works recursively.

File: tools/validate_spec.py [hunk 10]
```diff
@@ -161,6 +188,74 @@ def _validate_message(
...
+        self.target_coverage = target_coverage
+        for package in packages:
+            for message in package:
+                assert isinstance(message, MessageValue)
+                self._total_message_coverage[str(message.identifier)] = {
+                    link: False for link in message.model.structure
+                }
+
+                assert isinstance(message.model.location, Location)
+                assert isinstance(message.model.location.source, Path)
```
[the remark this answers is not in the document]
Comment (rssen) [reply tools/validate_spec.py]: They are unnecessary, but without them, i get mypy errors. Changed the instance check to an `is not None` check in 84cb4bf0.

File: tools/validate_spec.py [hunk 11]
```diff
@@ -161,6 +188,74 @@ def _validate_message(
...
+                file_name = message.model.location.source.name
+                if file_name in self.spec_files:
+                    self.spec_files[file_name].append(str(message.identifier))
+                else:
+                    self.spec_files[file_name] = [str(message.identifier)]
```
[the remark this answers is not in the document]
Comment (rssen) [reply tools/validate_spec.py]: Didn't know about `defaultdict` as of now. Thanks for mentioning it. Changed in 84cb4bf0

File: tools/validate_spec.py [hunk 12]
```diff
@@ -161,6 +188,74 @@ def _validate_message(
...
 
+class CoverageInformation:
+    def __init__(self, packages: List[Package], target_coverage: float) -> None:
+        self._total_message_coverage: Dict[str, Dict[Link, bool]] = {}
+        self.spec_files: Dict[str, List[str]] = {}
+        self.target_coverage = target_coverage
+        for package in packages:
+            for message in package:
+                assert isinstance(message, MessageValue)
+                self._total_message_coverage[str(message.identifier)] = {
```
[the remark this answers is not in the document]
Comment (rssen) [reply tools/validate_spec.py]: Because... I don't know... . Changed in 84cb4bf0.

File: tools/validate_spec.py [hunk 13]
```diff
@@ -226,10 +323,67 @@ def __exit__(
...
+        )
+        print("-" * 80)
+
+    @staticmethod
+    def __write_coverage_full(total_coverage: CoverageInformation) -> None:
+        print("\n")
+        print("=" * 80)
+        print(f"{'Link Coverage' : ^80}")
+        print("=" * 80)
+        for file, _ in total_coverage.spec_files.items():
```
[the remark this answers is not in the document]
Comment (rssen) [reply tools/validate_spec.py]: Fixed in 84cb4bf0

File: tools/validate_spec.py [hunk 14]
```diff
@@ -226,10 +323,67 @@ def __exit__(
...
+            print("\n")
+            print(f"{file : ^80}")
+            print("-" * 80)
+            uncovered_links = total_coverage.file_uncovered_links(file)
+            if len(uncovered_links) == 0:
+                print(f"{'all links covered':^80}")
+                continue
```
[the remark this answers is not in the document]
Comment (rssen) [reply tools/validate_spec.py]: Ok, changed in 815cc77c

File: tools/validate_spec.py (hunk 1 above)
Comment (rssen) [reply tools/validate_spec.py]: Ok, changed in 815cc77c

File: tools/validate_spec.py [hunk 15]
```diff
@@ -161,6 +188,74 @@ def _validate_message(
...
+        assert file_name in self.spec_files
+        uncovered = []
+        for message in self.spec_files[file_name]:
+            uncovered.extend(
+                [
+                    link
+                    for link, covered in self._total_message_coverage[message].items()
+                    if not covered
+                ]
+            )
```
[the remark this answers is not in the document]
Comment (rssen) [reply tools/validate_spec.py]: Changed in 398135ac. I'm not convinced that this change improves readability.

File: tests/test_validation_tool.py [hunk 16]
```diff
@@ -349,3 +349,47 @@ def test_validation_negative_full_output(tmp_path: Path) -> None:
...
+                "tests/data/ethernet/frame/invalid",
+                "-v",
+                "tests/data/ethernet/frame/valid",
+                "-c",
+                "--target-coverage=0.8",
+                "--no-verification",
+            ]
+        )
+        == "missed target coverage of 80.00%, reached 22.06%"
+    )
```
[the remark this answers is not in the document]
Comment (rssen) [reply tests/test_validation_tool.py]: Created a test data directory in f8ffc5d3

File: tests/test_validation_tool.py [hunk 17]
```diff
@@ -349,3 +349,47 @@ def test_validation_negative_full_output(tmp_path: Path) -> None:
...
+                "tests/data/ipv4/packet/valid",
+                "-i",
+                "tests/data/ipv4/packet/invalid",
+                "--coverage",
+                "--target-coverage=1",
+                "--no-verification",
+            ]
+        )
+        == 0
+    )
```
[the remark this answers is not in the document]
Comment (rssen) [reply tests/test_validation_tool.py]: Test added in 1e8aad0f

File: tests/test_validation_tool.py [hunk 18]
```diff
@@ -349,3 +349,47 @@ def test_validation_negative_full_output(tmp_path: Path) -> None:
         == f"{number} messages were classified incorrectly"
     )
     assert cmp(f"{tmp_path}/output.json", "tests/data/valid_full_output_negative.json")
+
+
+def test_validation_coverage() -> None:
```
[the remark this answers is not in the document]
Comment (rssen) [reply tests/test_validation_tool.py]: I don't understand this comment, did you expect a different test result?

File: tools/validate_spec.py [hunk 19]
```diff
@@ -226,10 +323,67 @@ def __exit__(
...
+            err = ValidationError()
+            if self.classified_incorrectly != 0:
+                err.append(f"{self.classified_incorrectly} messages were classified incorrectly")
+            if self.reached_coverage < self.target_coverage:
+                err.append(
+                    f"missed target coverage of {self.target_coverage:.2%}, "
+                    f"reached {self.reached_coverage:.2%}"
+                )
+            if len(err.messages) > 0:
+                raise err
```
[the remark this answers is not in the document]
Comment (rssen) [reply tools/validate_spec.py]: Don't know if it's a good reason, but I wanted to circumvent the explicit appending of a newline to each error message string, depending on whether the first, the second or both errors are raised.

File: tests/test_specs.py (hunk 3 above)
Comment (rssen) [reply tests/test_specs.py]: Done in f8ffc5d3.

File: tests/test_validation_tool.py [hunk 20]
```diff
@@ -349,3 +349,47 @@ def test_validation_negative_full_output(tmp_path: Path) -> None:
...
+                "-i",
+                "tests/data/ethernet/frame/invalid",
+                "-v",
+                "tests/data/ethernet/frame/valid",
+                "-c",
+                "--target-coverage=0.8",
+                "--no-verification",
+            ]
+        )
+        == "missed target coverage of 80.00%, reached 22.06%"
```
[the remark this answers is not in the document]
Comment (rssen) [reply tests/test_validation_tool.py]: Done in f8ffc5d3

File: tools/validate_spec.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (rssen) [reply tools/validate_spec.py]: Changed in f8ffc5d3

File: tools/validate_spec.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (treiher) [reply tools/validate_spec.py]: I don't think that is a good reason to make an exception class more complicated. Using a list and join here is simpler than adding that functionality to `ValidationError`.

File: tests/test_validation_tool.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (treiher) [reply tests/test_validation_tool.py]: Do we really need all of that specifications for testing the tool (e.g., ARP)? Does it make sense to name the test directory for the validation tool `coverage`?

File: tests/test_validation_tool.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (rssen) [reply tests/test_validation_tool.py]: Added in 433e961

File: tests/test_validation_tool.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (rssen) [reply tests/test_validation_tool.py]: ARP is a refinement for ethernet. So if we test `in_ethernet` we need the ARP specification.

File: tools/validate_spec.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (rssen) [reply tools/validate_spec.py]: Done in af3e3daa

File: tests/test_validation_tool.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (treiher) [reply tests/test_validation_tool.py]: > ARP is a refinement for ethernet. So if we test `in_ethernet` we need the ARP specification.

Do we really need all these refinements for testing the validation tool? Wouldn't it be sufficient to have just Ethernet, IPv4 and a refinement for both in In_Ethernet?

File: tools/validate_spec.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (treiher) [reply tools/validate_spec.py]: It seems I missed that in the last PR: The `ValidationError` does not need an implementation at all. The following is sufficient:

```Python
class ValidationError(Exception):
    pass
```

File: tools/validate_spec.py [hunk 21]
```diff
@@ -161,6 +215,112 @@ def _validate_message(
...
+            f"{'TOTAL' : <40} {self.total_links: >10} {self.total_covered_links : >10} "
+            f"{self.total_covered_links / self.total_links :15.2%}"
+        )
+        print("-" * 80)
+
+    def __print_link_coverage(self) -> None:
+        print("\n")
+        print("=" * 80)
+        print(f"{'Missing Link Coverage' : ^80}")
+        print("=" * 80)
```
Comment (treiher) [inline tools/validate_spec.py:310]: What is a "missing link coverage"? Did you mean "uncovered links"?

File: tools/validate_spec.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (treiher) [reply tools/validate_spec.py]: Yes, that solution is worse and not "just one list comprehension".

[the remark this answers is not in the document]
Comment (treiher) [reply tools/validate_spec.py]:
```diff
diff --git a/tools/validate_spec.py b/tools/validate_spec.py
index 19ca88d..9149eb7 100755
--- a/tools/validate_spec.py
+++ b/tools/validate_spec.py
@@ -263,18 +263,12 @@ class CoverageInformation:
 
     def file_uncovered_links(self, file_name: str) -> List[Link]:
         assert file_name in self.__spec_files
-        return list(
-            itertools.chain(
-                *[
-                    [
-                        link
-                        for link, covered in self.__total_message_coverage[message].items()
-                        if not covered
-                    ]
-                    for message in self.__spec_files[file_name]
-                ]
-            )
-        )
+        return [
+            link
+            for message in self.__spec_files[file_name]
+            for link, covered in self.__total_message_coverage[message].items()
+            if not covered
+        ]
 
     def print_coverage(self) -> None:
         if self.__coverage:
```

File: tests/test_validation_tool.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (treiher) [reply tests/test_validation_tool.py]: My first intuition was that misclassified messages should not count in the coverage, but after thinking about it, it is probably fine to keep both things independent from each other.

File: tools/validate_spec.py [hunk 22]
```diff
@@ -161,6 +215,112 @@ def _validate_message(
     )
 
 
+class CoverageInformation:
+    def __init__(self, packages: List[Package], coverage: bool) -> None:
+        self.__total_message_coverage: Dict[ID, Dict[Link, bool]] = {}
+        self.__spec_files: Dict[str, List[ID]] = defaultdict(lambda: [])
```
Comment (treiher) [inline tools/validate_spec.py:221]: Instead of `lambda: []` just `list` can be used.

File: tools/validate_spec.py [hunk 23]
```diff
@@ -161,6 +215,112 @@ def _validate_message(
...
+                for message in package:
+                    assert isinstance(message, MessageValue)
+                    self.__total_message_coverage[message.identifier] = {
+                        link: False for link in message.model.structure
+                    }
+
+                    assert message.model.location is not None
+                    assert message.model.location.source is not None
+                    file_name = message.model.location.source.name
+                    self.__spec_files[file_name].append(message.identifier)
```
Comment (treiher) [inline tools/validate_spec.py:235]: It is usually better to do essential checks very early in a function and just return:

```Python
if not self.__coverage:
    return
```

File: tools/validate_spec.py [hunk 24]
```diff
@@ -226,28 +398,10 @@ def __exit__(
...
 
-        if exception_value is None and self.classified_incorrectly != 0:
-            raise ValidationError(
-                f"{self.classified_incorrectly} messages were classified incorrectly"
-            )
-
     def write_result(self, validation_result: ValidationResult) -> None:
-        self.__write_console_output(validation_result)
+        validation_result.print_console_output()
         self.__write_json_output(validation_result)
```
Comment (treiher) [inline tools/validate_spec.py:403]: Why is `validation_result.print_console_output` called inside a method of `OutputWriter`?

File: tools/validate_spec.py [hunk 25]
```diff
@@ -1,16 +1,20 @@
 #!/usr/bin/env -S python3 -O
```
Comment (treiher) [inline tools/validate_spec.py:1]: Why is this one test so extremely slow?

```
53.75s call     tests/test_validation_tool.py::test_cli_not_regular_file
7.24s call     tests/test_validation_tool.py::test_validation_negative_full_output
6.73s call     tests/test_validation_tool.py::test_coverage_threshold_missed
6.65s call     tests/test_validation_tool.py::test_validation_positive_full_output
```

**Review by treiher — CHANGES_REQUESTED** [review]

File: tools/validate_spec.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (rssen) [reply tools/validate_spec.py]: Done in a8d48ed5

File: tools/validate_spec.py [hunk 26]
```diff
@@ -64,6 +68,21 @@ def cli(argv: List[str]) -> Union[int, str]:
...
+        action="store_true",
+        help="print the combined link-coverage of all provided messages",
+    )
+
+    parser.add_argument(
+        "--target-coverage",
+        type=float,
+        default=0,
+        help="abort with exitcode 1 if the coverage threshold is not reached, "
+        "target coverage expected in percentage",
```
[the remark this answers is not in the document]
Comment (rssen) [reply tools/validate_spec.py]: In 2a6c290c I changed it to `target coverage is expected in percentage`. That sounds better to my ears.

File: tools/validate_spec.py [hunk 27]
```diff
@@ -64,6 +68,21 @@ def cli(argv: List[str]) -> Union[int, str]:
         ),
     )
 
+    parser.add_argument(
+        "-c",
+        "--coverage",
+        action="store_true",
+        help="print the combined link-coverage of all provided messages",
```
[the remark this answers is not in the document]
Comment (rssen) [reply tools/validate_spec.py]: True. Changed it in 2a6c290c

File: tools/validate_spec.py [hunk 28]
```diff
@@ -122,10 +159,27 @@ def validate(
...
+        coverage
+        and coverage_info.total_covered_links / coverage_info.total_links < target_coverage / 100
+    ):
+        error_msgs.append(
+            f"missed target coverage of {target_coverage/100:.2%}, "
+            f"reached {coverage_info.total_covered_links / coverage_info.total_links:.2%}"
+        )
+    if len(error_msgs) > 0:
+        raise ValidationError("\n".join(e for e in error_msgs))
     return 0
```
[the remark this answers is not in the document]
Comment (rssen) [reply tools/validate_spec.py]: Removed the return value in e020204d

File: tools/validate_spec.py (hunk 22 above)
Comment (rssen) [reply tools/validate_spec.py]: Done in e020204d

File: tools/validate_spec.py (hunk 23 above)
Comment (rssen) [reply tools/validate_spec.py]: Done in e020204d

File: tools/validate_spec.py (hunk 24 above)
Comment (rssen) [reply tools/validate_spec.py]: Moved the call into `validate` in e020204d

File: tools/validate_spec.py (hunk 25 above)
Comment (rssen) [reply tools/validate_spec.py]: For this test, the `--no-verification` flag was missing. Fixed in 2072060a.

File: tools/validate_spec.py (hunk 21 above)
Comment (rssen) [reply tools/validate_spec.py]: I did not like "Uncovered Links" because it sounds like someone "discovered" the links by accident. Probably that's just me. Changed it in 19aca4c5.

File: tests/test_validation_tool.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (rssen) [reply tests/test_validation_tool.py]: I reduced the specs to IPv4, Ethernet, In_Ethernet and IANA (because it used in IPv4 and I wanted to stick to the original spec). 5a2b43c0

File: tools/validate_spec.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (rssen) [reply tools/validate_spec.py]: Changed in 7667713d

File: tests/test_validation_tool.py [hunk 29]
```diff
@@ -326,26 +327,78 @@ def test_validation_negative() -> None:
...
+        )
+        == 0
+    )
+    captured_output = capsys.readouterr()
+    with open("tests/validation_tool/valid_cov_output_ethernet.txt", "r") as f:
+        valid_output = f.read().replace("Directory: .", f"Directory: {os.getcwd()}")
+    assert captured_output.out == valid_output
+
+
+def test_coverage_threshold_missed(capsys: Any) -> None:
```
[the remark this answers is not in the document]
Comment (rssen) [reply tests/test_validation_tool.py]: If I use `CaptureFixture[str]` and import `CaptureFixture` from ` _pytest.capture` i get an `unused_import` error from pylint. If I remove the import (obviously) `CaptureFixture` is undefined. Additionaly I have to use `type: ignore` for that import becasue `_pytest.cypture` does not use type hints.
How can I supress mypy and pylint on the same line where the import is? Am I using the right import?

I have to put `CaptureFixture[str]` in quotes, because when used without, the error `CaptureFixture is not subscriptable` is raised.

[the remark this answers is not in the document]
Comment (treiher) [reply tests/test_validation_tool.py]: I cannot reproduce the issue. It is probably caused by an old version of pytest (or mypy/pylint). What version do you have? After the following changes all checks are successful for me:

```diff
diff --git a/tests/test_validation_tool.py b/tests/test_validation_tool.py
index 93ec1e1..c93007b 100644
--- a/tests/test_validation_tool.py
+++ b/tests/test_validation_tool.py
@@ -5,7 +5,7 @@ from pathlib import Path
 from typing import Iterator
 
 import pytest
-from _pytest.capture import CaptureFixture  # type: ignore
+from _pytest.capture import CaptureFixture
 from rflx.pyrflx import PyRFLX
 
 from tools.validate_spec import _validate_message, cli
@@ -354,7 +354,7 @@ def test_validation_negative_full_output(tmp_path: Path) -> None:
     assert cmp(f"{tmp_path}/output.json", "tests/validation_tool/valid_full_output_negative.json")
 
 
-def test_validation_coverage(capsys: CaptureFixture) -> None:
+def test_validation_coverage(capsys: CaptureFixture[str]) -> None:
     assert (
         cli(
             [
@@ -380,7 +380,7 @@ def test_validation_coverage(capsys: CaptureFixture) -> None:
     assert captured_output.out == valid_output
 
 
-def test_coverage_threshold_missed(capsys: CaptureFixture) -> None:
+def test_coverage_threshold_missed(capsys: CaptureFixture[str]) -> None:
     assert (
         cli(
             [
```

File: tools/validate_spec.py (hunk 21 above)
Comment (treiher) [reply tools/validate_spec.py]: Actually I never thought about the different meaning of "uncovered" in this context. AFAICT "uncovered lines" is a usual term wrt. to test coverage, so I think it makes sense to use a similar term here.

**Review by treiher — CHANGES_REQUESTED** [review]
LGTM except of the open mypy issue.

File: tests/test_validation_tool.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (rssen) [reply tests/test_validation_tool.py]: I have `mypy 0.812` and `pylint 2.8.2` and `pytest 5.4.3`. `pytest` is an old version but is constraint to a version < 6 in `setup.py` in RecodFlux (I'm using the same venv for RecodFlux and RecordFlux-specifications). I am also using python 3.9.4

[the remark this answers is not in the document]
Comment (treiher) [reply tests/test_validation_tool.py]: IIRC type hints where introduced in `pytest 6`, so this should be the cause. We have the version constraint for pytest because of Componolit/RecordFlux#416. Please check if the issue still occurs for recent versions (of pytest and pyicontract-lint). If not, we should change the dependency constraint in RecordFlux (to `pytest >=6, <7`) to allow using type hints also in RecordFlux.

[the remark this answers is not in the document]
Comment (rssen) [reply tests/test_validation_tool.py]: It works now with `CaptureFixture[str]` after upgrading pytest. 5088b56

**Review by treiher — APPROVED** [review]
